amaranth_soc/csr/bus.py

```diff
@@ -171,10 +172,87 @@ def memory_map(self, memory_map):
+            assert isinstance(name, str)
+            assert overlaps is None or isinstance(overlaps, int) and overlaps >= 0
+            self.name     = name
+            self.shape    = Shape.cast(unsigned(width))
+            self.overlaps = overlaps
+            self._ranges  = set()
+            self._blocks  = 1
+            self._chunks  = None
+
+        def add(self, elem_range):
+            assert isinstance(elem_range, range)
+            self._ranges.add(elem_range)
+
+        def decode(self, addr, elem_range):
+            assert isinstance(addr, int) and isinstance(elem_range, range)
+            range_bits = log2_int(elem_range.stop - elem_range.start, need_pow2=False)
+            block_bits = log2_int(self._blocks)
+            misaligned = bool(elem_range.start % (2 ** range_bits))
+            mask = 2 ** (range_bits + block_bits + misaligned) - 1
+            return addr & mask
+
+        def encode(self, offset, elem_range):
+            assert isinstance(offset, int) and isinstance(elem_range, range)
+            range_bits = log2_int(elem_range.stop - elem_range.start, need_pow2=False)
+            block_bits = log2_int(self._blocks)
+            misaligned = bool(elem_range.start % (2 ** range_bits))
+            mask = 2 ** (range_bits + block_bits + misaligned) - 1
+            return elem_range.start & ~mask | offset
+
+        def prepare(self):
+            if self.overlaps is None:
+                self.overlaps = len(self._ranges)
+
+            elements = defaultdict(list)
+            balanced = True
+
+            for elem_range in self._ranges:
+                for chunk_addr in elem_range:
+                    chunk_offset = self.decode(chunk_addr, elem_range)
+                    if len(elements[chunk_offset]) > self.overlaps:
+                        balanced = False
+                        break
+                    elements[chunk_offset].append(elem_range)
+
+            if balanced:
+                self._ranges = frozenset(self._ranges)
+                self._chunks = dict()
+                for chunk_offset, chunk_elements in elements.items():
+                    chunk = Multiplexer._Shadow.Chunk(self, chunk_offset, chunk_elements)
+                    self._chunks[chunk_offset] = chunk
+            else:
+                self._blocks *= 2
+                self.prepare()
+
+        def chunks(self):
+            for chunk_offset, chunk in self._chunks.items():
+                yield chunk_offset, chunk
+
     """CSR register multiplexer.
 
     An address-based multiplexer for CSR registers implementing atomic updates.
 
+    This implementation assumes the following from the CSR bus:
+        * an initiator must have exclusive ownership over the multiplexer for the full duration of
+          a register transaction;
+        * an initiator must access a register in ascending order of addresses, but it may abort a
+          transaction after any bus cycle.
```

Comment:
Maybe add something like this?

```suggestion

          For large buses or buses with wide registers, the amount of resources consumed is significantly influenced by:
              * Alignment of CSR registers to their next power-of-2 boundary.
              * The ``shadow_overlaps`` parameter.
          These two factors are explained in detail below, and can be used to optimize the multiplexer's implementation.
```

(Please feel free to reword this, this is just my rough sketch.)

Reply:
Alignment no longer plays a significant role in resource consumption, it was design mistake. Fixing it removed the need to document it, while saving a few more resources.

Reply:
:+1: